ACPI: fix multi-page ACPI table allocation

wipawel opened the pull request:
An ACPI table may span multiple 4K pages because of its multi-page
size or its starting offset (page crossing table).
Introduce `acpi_table_map_pages()` and `acpi_table_unmap_pages()` routines
properly detecting starting page as well as number of pages required
for a table.

It is also essential to map entire ACPI table before its checksum
validation.  For almost all tables (except XSDT) the mapping process
needs to be split into ACPI table header page mapping, where the
entire table length can be obtained, and then the entire table
mapping.

Signed-off-by: Pawel Wieczorkiewicz <wipawel@amazon.de>

* Issue #102 *
* Fixes #102 *

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

bjoernd commented on common/acpi.c:
```diff
@@ -39,6 +39,11 @@ unsigned max_acpi_tables;
 
+unsigned acpi_get_nr_cpus(void) { return nr_cpus; }
+
+#define ACPI_NR_TABLES(ptr)                                                              \
```
How about adding a comment what the logic behind this calculation is? I'm not getting it on first read.

bjoernd reviewed (commented):
It seems this is merging a couple of commits into one, which makes reviewing more difficult than it should be.

wipawel committed:
acpi: minor cleanup

Move defines and one-line function to the top of the file.

Signed-off-by: Pawel Wieczorkiewicz <wipawel@amazon.de>

wipawel committed:
acpi: add acpi_dump_table() helper routine

Split acpi_dump_tables() routine into a single table helper and a loop
over it.  This is needed to dump RSDT/XSDT single tables without
redundant code.

Signed-off-by: Pawel Wieczorkiewicz <wipawel@amazon.de>

wipawel committed:
ACPI: fix multi-page ACPI table allocation

An ACPI table may span multiple 4K pages because of its multi-page
size or its starting offset (page crossing table).
Introduce acpi_table_map_pages() and acpi_table_unmap_pages() routines
properly detecting starting page as well as number of pages required
for a table.

It is also essential to map entire ACPI table before its checksum
validation.  For almost all tables (except XSDT) the mapping process
needs to be split into ACPI table header page mapping, where the
entire table length can be obtained, and then the entire table
mapping.

Signed-off-by: Pawel Wieczorkiewicz <wipawel@amazon.de>

wipawel committed:
Merge branch 'mainline' into acpi_fix
